presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java

```diff
@@ -44,11 +48,26 @@ public long getRetainedSizeInBytes()
 
     public Page removePage()
     {
-        int referenceCount = this.referenceCount.decrementAndGet();
+        int referenceCount = REFERENCE_COUNT_UPDATER.decrementAndGet(this);
         checkArgument(referenceCount >= 0, "Page reference count is negative");
         if (referenceCount == 0) {
-            onFree.run();
+            onPageReleased.accept(page.getRetainedSizeInBytes());
         }
         return page;
     }
+
+    @Override
```

Comment:
extract to separate commit

Reply:
This was a bit of a hack, without adding this toString() implementation the `referenceCount` field is detected as unused and fails the build.

Reply:
You could add `toString` without `referenceCount` in previous commit and then add `referenceCount` to `toString` in this commit